src/Stratis.FederatedPeg.Features.FederationGateway/SourceChain/MaturedBlocksProvider.cs

```diff
@@ -50,20 +54,51 @@ public async Task<List<MaturedBlockDepositsModel>> GetMaturedDepositsAsync(int b
                 throw new InvalidOperationException($"Block height {blockHeight} submitted is not mature enough. Blocks less than a height of {matureTipHeight} can be processed.");
             }
 
+            // Cache clean-up.
+            lock (this.depositCache)
+            {
+                // The requested height gives away the fact that the peer is probably no longer interested in cached entries below that height.
+                // Keep an additional 1,000 blocks anyway in case there are some parallel request that are still executing for lower heights.
+                foreach (int i in this.depositCache.Where(d => d.Key < (blockHeight - 1000)).Select(d => d.Key).ToArray())
+                    this.depositCache.Remove(i);
+            }
+
             var maturedBlocks = new List<MaturedBlockDepositsModel>();
 
+            // Don't spend to much time that the requester may give up.
+            DateTime deadLine = DateTime.Now.AddSeconds(30);
```

Comment:
Might be an idea to tie this directly to the client timeout in some way, but this should be fine for now.